src/algorithms/calorimetry/TrackClusterMergeSplitter.cc

```diff
@@ -0,0 +1,460 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2024 Derek Anderson
+
+#define EICRECON_TRACKCLUSTERMERGESPLITTER_CC
+
+#include <cmath>
+#include <regex>
+#include <utility>
+#include <algorithm>
+// dd4hep utilities
+#include <DD4hep/Readout.h>
+// jana utilities
+#include <JANA/JException.h>
+// edm4hep types
+#include <edm4hep/Vector3f.h>
+#include <edm4hep/utils/vector_utils.h>
+// edm4eic types
+#include <edm4eic/ProtoCluster.h>
+#include <edm4eic/TrackSegment.h>
+#include <edm4eic/CalorimeterHit.h>
+
+// algorithm definition
+#include "TrackClusterMergeSplitter.h"
+
+
+
+namespace eicrecon {
+
+  // --------------------------------------------------------------------------
+  //! Initialize algorithm
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::init(
+    const dd4hep::Detector* detector,
+    const dd4hep::rec::CellIDPositionConverter* converter
+  ) {
+
+    m_detector = detector;
+    m_converter = converter;
+
+  }  // end 'init(dd4hep::Detector*)'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Process inputs
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::process(
+    const TrackClusterMergeSplitter::Input& input,
+    const TrackClusterMergeSplitter::Output& output
+  ) const {
+
+    // grab inputs/outputs
+    const auto [in_protoclusters, in_projections] = input;
+    auto [out_protoclusters] = output;
+
+    // exit if no clusters in collection
+    if (in_protoclusters -> size() == 0) {
+      throw JException("No proto-clusters in input collection!");
+    }
+
+    // reset bookkeeping containers
+    reset_bookkeepers();
+
+    // flag all clusters as not being consumed
+    for (auto cluster : *in_protoclusters) {
+      m_mapIsConsumed.insert(
+        {cluster.getObjectID().index, false}
```

Comment:
Using just the index in the collection without the full id will break if input collection is a subset collection combining several other inputs. Is it possible to use full ObjectID's?

Reply:
Ah! That's something I hadn't considered! It should be, but I might need to rework how I'm doing my bookkeeping (which might be a good thing since I can probably simplify things)...

Reply:
Thinking through this: would subsets be an issue if rather than using `ObjectID().index` I used the index of the cluster/projection in the input collections, e.g. `iCluster` in (*in_protoclusters)[iCluster]`?

Reply:
It's the same value, no?

Reply:
I guess, it's not for the subset collections. Not sure why not just use ObjectID, but this would be fine too.

Reply:
Gotcha! I've been working with just the indices since there are a few times where I retrieve an object from the input using the index (e.g. at [line 181](https://github.com/eic/EICrecon/blob/a72118deb86a22368b9f445425b37b67881a74d0/src/algorithms/calorimetry/TrackClusterMergeSplitter.cc#L181C1-L186C69)).:
```
      // add clusters to be merged
      vecClust.clear();
      for (const int& iClustToMerge : clustToMerge.second) {
        vecClust.push_back( (*in_protoclusters)[iClustToMerge] );
      }
      vecClust.push_back( (*in_protoclusters)[clustToMerge.first] );
```

I definitely could key the maps on ObjectID's, and then use `ObjectID().index` when needed... But I would think that would still run into the subset issue, right?

Reply:
That would not, as object id also contains collection id.

Reply:
Gotcha! I'll just do that, then!